### DISCUSSION TITLE
Write a tiff file tile-by-tile

### DISCUSSION THREAD

**Discussion by jonasteuwen** [libvips/libvips#2860] [Q&A]
Currently I am writing the output of a CNN to a TIFF by using:

```python
def from_iterator(self, iterator: Iterable, save_path: Union[str, os.PathLike], total: Optional[int] = None):
        vips_image = None
        for tile_index, (tile_coordinates, _tile) in tqdm(
            enumerate(iterator), disable=self._silent, unit="tiles", total=total
        ):
            _tile = np.asarray(_tile)
            if vips_image is None:
                # Assumes last axis is the channel!
                vips_image = pyvips.Image.black(*self._size, bands=_tile.shape[-1])

            vips_tile = numpy_to_vips(_tile)
            vips_image = vips_image.insert(vips_tile, *tile_coordinates)

        _save_tiff(vips_image, save_path)
```

This of course is slightly problematic because it first assembles the image and then writes it to tiff. Would there be a way by using libvips itself and write this tile-by-tile to the image? Could use pybind11 to make a writer that way. `_save_tiff` is the pyvips tiffwriter.

**Comment by jcupitt** [comment]
Hi @jonasteuwen,

Sadly, not using pyvips. You can do it in C/C++, but that's probably no help.

We've been talking about modifying libvips and pyvips to make this possible, but no one's found the time yet.

**Reply by jonasteuwen** [reply]
Hi @jcupitt, yes, C++ would actually be fine, I can wrap it with pybind11. Would you have some pointers how to do that? As I see it the tiff writer takes a full image as well.

**Reply by jcupitt** [reply]
You'd need to write your own vips operator (subclass VipsOperator, implement `_generate()`) and in the generate method, call off into Python, perhaps with some locking to manage concurrency. There'd probably be some tricky issues to resolve around initialisation.

**Reply by lromor** [reply]
Hi @jcupitt, thank you for your answer, could you provide a bit more context? I guess you are referring to https://www.libvips.org/API/current/VipsOperation.html right? So a user could define a vips operation that performs immediate tile writing and call this operator from python. Is that correct? How would you see that happening in pseudocode with the final operator?

**Reply by jcupitt** [reply]
The idea would be to present your tile generator as a vipsimage. In Python, you'd then do something like:

```python
def make_a_tile(rect):
    # compute an (x, y, w, h) rect of pixels
    return a_tile

# make a 100,000 x 100,000 image and use make_a_tile as the function
# that's used to compute tiles on demand
big_image = pyvips.Image.my_tile_generator(100000, 100000, make_a_tile)

save_tiff(big_image, save_path)
```

Where `my_tile_generator` is the subclass of VipsOperation you'd need to write. You'd then have to hook your CNN up so it could produce tiles on request. You can arrange the pipeline so that tiles are only requested in a fixed size, and only in top-to-bottom order, if that helps.
